### PR TITLE
Minor changes in Slice GPU kernels, before reusing them in NumpyReader GPU

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

#### Why we need this PR?
*Pick one, remove the rest*
- Refactoring to improve reusability of slice GPU kernels (SliceGPU, SliceFlipNormalizePermutePadGpu)

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     *Fixed issues when using the above mentioned kernels in more compilation units.:*
     *Collision of BlockDesc name from SliceGPU and SliceFlipNormalizePermutePadGpu*
     *Special __ldg case for bool*
     *Special case for empty cropping window*
 - Affected modules and functionalities:
     *Slice family operators*
 - Key points relevant for the review:
     *NA*
 - Validation and testing:
     *Existing tests apply*
 - Documentation (including examples):
     *NA*

**JIRA TASK**: *[DALI-2127]*

### CODE REVIEW

**Comment by jantonguirao** [comment]
!build

File: dali/kernels/slice/slice_gpu.cuh [hunk 1]
```diff
@@ -86,11 +97,15 @@ __device__ void SliceFuncNoPad(OutputType *__restrict__ out, const InputType *__
...
-DALI_HOST_DEV DALI_FORCEINLINE bool is_out_of_bounds(int64_t idx, int64_t data_extent) {
-  // check idx < 0 and idx >= data_extent at once
-  return static_cast<uint64_t>(idx) >= static_cast<uint64_t>(data_extent);
+template <typename T>
+DALI_HOST_DEV DALI_FORCEINLINE T ldg(const T* ptr) {
+  return __ldg(ptr);
 }
 
+template <>
+DALI_HOST_DEV DALI_FORCEINLINE bool ldg(const bool* ptr) {
```
Comment (mzient) [inline dali/kernels/slice/slice_gpu.cuh:106]: Is there no __ldg for bool? If so:
```suggestion
DALI_HOST_DEV DALI_FORCEINLINE bool ldg(const bool* ptr) {
  return __ldg(reinterpret_cast<const type_of_size<sizeof(bool)> *>);
}
```
`type_of_size` is in `kernels/common/type_erasure.h`

File: dali/kernels/slice/slice_gpu.cuh [hunk 2]
```diff
@@ -86,11 +97,15 @@ __device__ void SliceFuncNoPad(OutputType *__restrict__ out, const InputType *__
...
+  return *ptr;
+}
```
Comment (mzient) [inline dali/kernels/slice/slice_gpu.cuh]: Is there no __ldg for bool? If so:
```suggestion
__device__ DALI_FORCEINLINE bool __ldg(const bool* ptr) {
  return __ldg(reinterpret_cast<const type_of_size<sizeof(bool)> *>(ptr));
}
```
`type_of_size` is in `kernels/common/type_erasure.h`

File: dali/kernels/slice/slice_gpu.cuh [hunk 3]
```diff
@@ -151,12 +166,13 @@ __device__ void SliceFunc(OutputType *__restrict__ out, const InputType *__restr
       in_idx += i_d;  // in_strides[d] is 1
 
     // Fill values are reused a lot, so let's make sure they are cached (by using __ldg())
-    out[out_idx] = out_of_bounds ? __ldg(&fill_values[i_c]) : clamp<OutputType>(in[in_idx]);
+    out[out_idx] =
+        out_of_bounds ? ldg<OutputType>(&fill_values[i_c]) : clamp<OutputType>(in[in_idx]);
```
Comment (mzient) [inline dali/kernels/slice/slice_gpu.cuh:171]: Revert once the overload is there.

File: dali/kernels/slice/slice_gpu.cuh [hunk 4]
```diff
@@ -86,11 +97,15 @@ __device__ void SliceFuncNoPad(OutputType *__restrict__ out, const InputType *__
   }
 }
 
-DALI_HOST_DEV DALI_FORCEINLINE bool is_out_of_bounds(int64_t idx, int64_t data_extent) {
-  // check idx < 0 and idx >= data_extent at once
-  return static_cast<uint64_t>(idx) >= static_cast<uint64_t>(data_extent);
+template <typename T>
+DALI_HOST_DEV DALI_FORCEINLINE T ldg(const T* ptr) {
+  return __ldg(ptr);
```
Comment (mzient) [inline dali/kernels/slice/slice_gpu.cuh]: Remove - this isn't necvessary.

File: dali/kernels/slice/slice_gpu.cuh [hunk 5]
```diff
@@ -151,12 +161,13 @@ __device__ void SliceFunc(OutputType *__restrict__ out, const InputType *__restr
       in_idx += i_d;  // in_strides[d] is 1
 
     // Fill values are reused a lot, so let's make sure they are cached (by using __ldg())
-    out[out_idx] = out_of_bounds ? __ldg(&fill_values[i_c]) : clamp<OutputType>(in[in_idx]);
+    out[out_idx] =
+        out_of_bounds ? __ldg(&fill_values[i_c]) : clamp<OutputType>(in[in_idx]);
```
Comment (mzient) [inline dali/kernels/slice/slice_gpu.cuh:165]: Nitpick: don't reformat if there are no changes and the formatting isn't obviously broken.

**Review by mzient — APPROVED** [review]

File: dali/kernels/slice/slice_gpu.cuh (hunk 5 above)
Comment (jantonguirao) [reply dali/kernels/slice/slice_gpu.cuh:165]: Good point, it was a leftover. I'll revert

**Review by awolant — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

**Pull request merged**